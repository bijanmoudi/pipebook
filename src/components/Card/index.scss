/*===================================
=            Card Styles            =
===================================*/

@import '../../assets/styles/constants/layout';
@import '../../assets/styles/helpers/colors';
@import '../../assets/styles/helpers/units';
@import '../../../node_modules/sass-mq/mq';

.card {
	position: relative;
	border-radius: rem($border-radius);
	border: 1px solid get-color('gray-light-very');
	list-style: none;
	box-shadow: 0 0 rem(3) get-color('gray-light', 0.45);
	background-color: get-color('white', 0.9);
	animation: fadeIn 500ms;
	&:hover {
		box-shadow: 0 0 rem(3) get-color('gray-dark-extra', 1);
		.card {
			&__header-extra {
				opacity: 1;
				visibility: visible;
				transform: scale(1);
			}
		}
	}
	&:not(:hover) {
		.card {
			&__header-extra {
				@media (hover: hover) and (pointer: fine) {
					border: 0;
					clip: rect(0 0 0 0);
					height: 1px;
					margin: -1px;
					overflow: hidden;
					padding: 0;
					position: absolute;
					white-space: nowrap;
					width: 1px;
				}
			}
		}
	}
	&.is--draggable {
		user-select: none;
		.card__handle {
			cursor: move;
			cursor: grab;
			cursor: -moz-grab;
			cursor: -webkit-grab;
			&:active {
				cursor: grabbing;
				cursor: -moz-grabbing;
				cursor: -webkit-grabbing;
			}
		}
	}
	& + & {
		margin-top: em(10);
	}
	&__handle {
		z-index: 0;
		position: absolute;
		top: 0;
		bottom: 0;
		left: 0;
		right: 0;
	}
	&__inner {
		display: flex;
		flex-direction: row-reverse;
		align-items: center;
		margin: 0 em(-5);
		padding: em(14) em(30);
		@include mq($until: sm) {
			flex-direction: column;
			text-align: center;
		}
	}
	&__column {
		padding: em(4) em(5);
		&--info {
			flex-grow: 1;
		}
	}
	&__header {
		min-height: em(22);
		& > * {
			z-index: 1;
			position: relative;
			display: inline-block;
			vertical-align: middle;
			margin-right: em(10);
			@include mq($until: sm) {
				display: inline;
				margin-right: 0;
			}
		}
		& + * {
			margin-top: em(3);
		}
	}
	&__title {
		color: get-color('gray-dark-very');
		& > * {
			display: block;
			@include mq($from: sm) {
				max-width: em(420);
				white-space: nowrap;
				text-overflow: ellipsis;
				overflow: hidden;
			}
			@include mq($until: sm) {
				display: inline;
			}
		}
		a {
			color: inherit;
			text-decoration: none;
			&:hover,
			&:active,
			&:focus {
				text-decoration: underline;
			}
			&:active {
				transform: scale(1);
				opacity: 0.8;
			}
		}
	}
	&__header-extra {
		line-height: 1;
		@media (hover: hover) and (pointer: fine) {
			visibility: hidden;
			opacity: 0;
			transform: scale(0);
			transition: opacity 250ms ease, visibility 250ms ease,
				transform 250ms ease;
		}
		@include mq($until: sm) {
			position: absolute;
			top: em(10);
			left: em(10);
		}
	}
	&__meta {
		color: get-color('gray');
		line-height: 1;
		&-item {
			z-index: 1;
			position: relative;
			display: inline-flex;
			margin-right: em(12);
			text-align: left;
			font-size: em(14);
			@include mq($until: sm) {
				margin: 0 em(6);
			}
			&--organization {
				&:before {
					content: '';
					flex-shrink: 0;
					width: em(15);
					height: em(15);
					margin-right: em(6, 14);
					background: url('../../assets/images/icons/office.svg') center center
						no-repeat;
				}
			}
		}
	}
	&__avatar {
		z-index: 1;
		position: relative;
		@include mq($until: sm) {
			font-size: em(29);
		}
	}
}
