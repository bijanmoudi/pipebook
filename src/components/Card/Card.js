import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import stringReplace from 'react-string-replace';
import { Avatar } from '../../components';
import './index.scss';

const Card = ({
	title,
	titleExtra,
	highlight,
	link,
	metadata,
	avatar,
	contentExtra,
	elementType,
	className,
	...restProps
}) => {
	const ElementType = elementType;
	const highlightedtitle = !highlight
		? title
		: stringReplace(title, highlight, (match, i) => (
				<span key={`__highlight${i}`} className="highlighted">
					{match}
				</span>
		  ));
	return (
		<ElementType
			className={classnames('card', { [className]: className })}
			{...restProps}
		>
			<div className="card__inner">
				<div className="card__column card__column--image">
					<Avatar
						className="card__avatar"
						name={title}
						url={avatar}
						link={link}
					/>
				</div>
				<div className="card__column card__column--info">
					<div className="card__header">
						<h3 className="card__title">
							{link ? (
								<a href={link}>{highlightedtitle}</a>
							) : (
								<span>{highlightedtitle}</span>
							)}
						</h3>
						{titleExtra && (
							<div className="card__header-extra">
								{titleExtra}
							</div>
						)}
					</div>
					{metadata && Object.keys(metadata).length ? (
						<ul className="card__meta">
							{Object.keys(metadata).map((meta, i) => (
								<li
									key={`__meta${i}`}
									className={classnames(
										'card__meta-item',
										`card__meta-item--${meta}`
									)}
								>
									<span>{metadata[meta]}</span>
								</li>
							))}
						</ul>
					) : null}
				</div>
			</div>
			{contentExtra}
		</ElementType>
	);
};

Card.propTypes = {
	title: PropTypes.string.isRequired,
	titleExtra: PropTypes.oneOfType([() => null, PropTypes.element]),
	highlight: PropTypes.oneOfType([() => null, PropTypes.string]),
	link: PropTypes.oneOfType([() => null, PropTypes.string]),
	metadata: PropTypes.oneOfType([() => null, PropTypes.object]),
	avatar: PropTypes.oneOfType([() => null, PropTypes.string]),
	contentExtra: PropTypes.oneOfType([() => null, PropTypes.element]),
	elementType: PropTypes.string
};

Card.defaultProps = {
	title: null,
	titleExtra: null,
	highlight: null,
	link: null,
	metadata: null,
	avatar: null,
	contentExtra: null,
	elementType: 'li'
};

export default Card;
