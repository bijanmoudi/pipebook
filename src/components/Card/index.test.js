import React from 'react';
import renderer from 'react-test-renderer';
import Card from './Card';

const props = {
	title: 'Dulce Mays',
	titleExtra: (
		<a href="#help" className="button">
			?
		</a>
	),
	highlight: 'Dul',
	link: 'https://www.google.com/search?q=Dulce+Mays',
	metadata: {
		organization: 'Rockford Linear Actuation'
	},
	avatar: 'http://placehold.it/200x200',
	contentExtra: <div className="card__handle" />,
	elementType: 'div'
};

describe('Testing "Card" component', () => {
	it('Rendering the default', () => {
		const tree = renderer
			.create(
				<Card
					title={props.title}
					titleExtra={props.titleExtra}
					link={props.link}
					avatar={props.avatar}
				/>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('Rendering with highlighted text', () => {
		const tree = renderer
			.create(
				<Card
					title={props.title}
					link={props.link}
					avatar={props.avatar}
					highlight={props.highlight}
					metadata={props.metadata}
				/>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('Rendering as a draggable "div"', () => {
		const tree = renderer
			.create(
				<Card
					title={props.title}
					link={props.link}
					avatar={props.avatar}
					contentExtra={props.contentExtra}
					elementType={props.elementType}
				/>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
