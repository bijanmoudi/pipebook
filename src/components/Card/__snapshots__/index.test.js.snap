// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing "Card" component Rendering as a draggable "div" 1`] = `
<div
  className="card"
>
  <div
    className="card__inner"
  >
    <div
      className="card__column card__column--image"
    >
      <figure
        className="avatar avatar--is-loading card__avatar"
      >
        <a
          className="avatar__wrapper"
          href="https://www.google.com/search?q=Dulce+Mays"
          tabIndex="-1"
          title="Dulce Mays"
        >
          <abbr
            className="avatar__abbreviation"
          >
            DM
          </abbr>
          <div
            aria-label="Dulce Mays's Avatar"
            className="avatar__image"
            role="img"
            style={
              Object {
                "backgroundImage": "url(null)",
              }
            }
          />
        </a>
      </figure>
    </div>
    <div
      className="card__column card__column--info"
    >
      <div
        className="card__header"
      >
        <h3
          className="card__title"
        >
          <a
            href="https://www.google.com/search?q=Dulce+Mays"
          >
            Dulce Mays
          </a>
        </h3>
      </div>
    </div>
  </div>
  <div
    className="card__handle"
  />
</div>
`;

exports[`Testing "Card" component Rendering the default 1`] = `
<li
  className="card"
>
  <div
    className="card__inner"
  >
    <div
      className="card__column card__column--image"
    >
      <figure
        className="avatar avatar--is-loading card__avatar"
      >
        <a
          className="avatar__wrapper"
          href="https://www.google.com/search?q=Dulce+Mays"
          tabIndex="-1"
          title="Dulce Mays"
        >
          <abbr
            className="avatar__abbreviation"
          >
            DM
          </abbr>
          <div
            aria-label="Dulce Mays's Avatar"
            className="avatar__image"
            role="img"
            style={
              Object {
                "backgroundImage": "url(null)",
              }
            }
          />
        </a>
      </figure>
    </div>
    <div
      className="card__column card__column--info"
    >
      <div
        className="card__header"
      >
        <h3
          className="card__title"
        >
          <a
            href="https://www.google.com/search?q=Dulce+Mays"
          >
            Dulce Mays
          </a>
        </h3>
        <div
          className="card__header-extra"
        >
          <a
            className="button"
            href="#help"
          >
            ?
          </a>
        </div>
      </div>
    </div>
  </div>
</li>
`;

exports[`Testing "Card" component Rendering with highlighted text 1`] = `
<li
  className="card"
>
  <div
    className="card__inner"
  >
    <div
      className="card__column card__column--image"
    >
      <figure
        className="avatar avatar--is-loading card__avatar"
      >
        <a
          className="avatar__wrapper"
          href="https://www.google.com/search?q=Dulce+Mays"
          tabIndex="-1"
          title="Dulce Mays"
        >
          <abbr
            className="avatar__abbreviation"
          >
            DM
          </abbr>
          <div
            aria-label="Dulce Mays's Avatar"
            className="avatar__image"
            role="img"
            style={
              Object {
                "backgroundImage": "url(null)",
              }
            }
          />
        </a>
      </figure>
    </div>
    <div
      className="card__column card__column--info"
    >
      <div
        className="card__header"
      >
        <h3
          className="card__title"
        >
          <a
            href="https://www.google.com/search?q=Dulce+Mays"
          >
            
            <span
              className="highlighted"
            >
              Dul
            </span>
            ce Mays
          </a>
        </h3>
      </div>
      <ul
        className="card__meta"
      >
        <li
          className="card__meta-item card__meta-item--organization"
        >
          <span>
            Rockford Linear Actuation
          </span>
        </li>
      </ul>
    </div>
  </div>
</li>
`;
